wellcomeml/datasets/winer.py

```diff
@@ -0,0 +1,222 @@
+                        if sentence_entities.get(sentence_id):
+                            sentence_entities[sentence_id].append([begin, end, ent_type])
+                        else:
+                            sentence_entities[sentence_id] = [[begin, end, ent_type]]
+                    entities[article_id] = sentence_entities
+
+    # Create id2word from vocab file (line number is id)
+    print("Creating id2word dictionary from {}".format(vocab_path))
+    id2word = {}
+    with open(vocab_path, "r") as vocab:
+        f = vocab.read()
+        lines = f.split('\n')
+        for i, line in enumerate(lines):
+            id2word[i] = line.split(' ')[0]
+
+    # Go into a sample of the document files and get the text for these entities
+    print("Creating token and tag from {}".format(docs_path))
+    with tarfile.open(docs_path, "r:bz2") as tar_docs, \
+            open(train_processed_path, 'w') as train_file, \
+            open(test_processed_path, 'w') as test_file:
+        for i, member in tqdm(enumerate(random.sample(tar_docs.getmembers(), n_sample))):
+            if i < n_train:
+                output_file = train_file
+            else:
+                output_file = test_file
+            f = tar_docs.extractfile(member)
+            if f:
+                content = f.read().decode('utf-8', errors='ignore')
+                articles = content.replace('\t', ' ').split('ID ')
+                if articles[0]=='':
+                    articles = articles[1:]
+                for article in articles:
+                    lines = article.split('\n')
+                    article_id = lines[0]
+                    article_entities = entities.get(article_id)
+                    if article_entities:
+                        output_file.write('ID ' + article_id)
+                        output_file.write('\n')
+                        if lines[-1]=='':
+                            lines = lines[1:-1]
+                        else:
+                            lines = lines[1:]
+                        sent_wordidx = [l.split(' ') for l in lines]
+                        for sentid, sentence_entities in article_entities.items():
+                            if len(sent_wordidx) > int(sentid):
+                                tokens = [id2word[int(s)] for s in sent_wordidx[int(sentid)]]
+                                tags = ['O']*len(tokens)
+                                for begin, end, ent_type in sentence_entities:
+                                    tags[(begin+1):(end-1)] = [str(ent_type) + '-I']*(end-1-begin-1)
+                                    tags[end-1] = str(ent_type) + '-E'
+                                    tags[begin] = str(ent_type) + '-B'
+                                for token, tag in list(zip(tokens, tags)):
+                                    output_file.write(' '.join([token, tag]))
+                                    output_file.write('\n')
+                        output_file.write('\n')
```

Comment:
I am not 100% sure of the logic, but it seems that is is outputting a double `\n` for the last entity?

Reply:
Ah yes, this should be indented, good spot! I want each sentence within a article to be separated by a newline, like
```
ID 123
The O
Liz PER
is O
. O

In O
Germany LOC
. O

ID 345
Another O
article O
```